README.md
# Deep Learning Handwriting Recognition

A simple python application that recognizes handwriting and converts it into text, using a multiple machine models (5). [EMNIST Dataset](https://www.kaggle.com/crawford/emnist) on Kaggle balance data set was used to train the ML. 
This can recognize digits, small and capital letters.

The models were trained on the following characters: `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabdefghnqrt`

## Things to do before running:
```commandline
mkdir dataset
mkdir generated_pickle
mkdir logs
```

## How to Run:
Train the dataset into a proper model
```commandline
Provide the train.csv, test.csv and mapping.txt.
Also check the train and test data if the orientation is correct, otherwise make necessary adjustments to rotate or flip or even convert it to black on white or vise versa.

RUN:
python3 train_dataset.py
```


Permutate the Model
I used 5 models to predict the input and run a voting system to get the final prediction.

model_1 and model_2:
ARCH: 2 convolutional layers, 2 dense layers, 128 neurons, 0.2 dropout, ran at 10 epochs.

model_3, model_4, and model_5:
ARCH: 2 convolutional layers, 2 dense layers, 64 neurons, 0.2 dropout, ran at 10 epochs.

```commandline
Change the dense_layers, layer_size and conv_layers to desired architecture

RUN:
python3 model_permutation.py
```




PREDICT images
Add additional or completly change images to predict in test/{number}.png
Make sure to use 28 x 28 images.
```
python3 predict.py
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v1.0.0] - 2022-03-13

### Added
    Initial Commit, testing 10 images in 28 x 28 dimension and predict in small cases.

    

test.py
import os
import tensorflow as tf
import numpy as np
import cv2
import matplotlib.pyplot as plt
from PIL import Image

TEST_CASE_SMALL = True
TEST_DIR_CAPITAL = 'test/capital'
TEST_DIR_SMALL = 'test/small'

TEST_DIR = TEST_DIR_CAPITAL
if TEST_CASE_SMALL:
    TEST_DIR = TEST_DIR_SMALL

# defining the models
model_1 = tf.keras.models.load_model('models/model_1.h5')  # 2 dense layer 2 convo layer 128 nodes
model_2 = tf.keras.models.load_model('models/model_2.h5')  # 2 dense layer 2 convo layer 128 nodes
model_3 = tf.keras.models.load_model('models/model_3.h5')  # 2 dense layer 2 convo layer 64 nodes
model_4 = tf.keras.models.load_model('models/model_4.h5')  # 2 dense layer 2 convo layer 64 nodes
model_5 = tf.keras.models.load_model('models/model_5.h5')  # 2 dense layer 2 convo layer 64 nodes

# defining mapping
class_mapping = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabdefghnqrt'
class_length = len(class_mapping)

IMG_SIZE = 28

def make_prediction(model, image):
    prediction = model.predict(image)
    idx_prediction = np.argmax(prediction[0])
    return class_mapping[idx_prediction]


def model_jury_ruling(*argv):
    all_predictions = []
    for arg in argv:
        all_predictions.append(arg)

    hash = {}

    for prediction in all_predictions:
        if prediction not in hash:
            hash[prediction] = 1
        else:
            hash[prediction] += 1

    return max(hash, key=hash.get)


def take_all_prediction(image):
    predictions = []
    models = [model_1, model_2, model_3, model_4, model_5]
    for model in models:
        predicted = make_prediction(model, image)
        predictions.append(predicted)
    return predictions


def prepare(test_number):
    image = cv2.imread(f"{TEST_DIR}/{test_number}.png", cv2.IMREAD_GRAYSCALE)
    image = np.invert(np.array([image]))
    array_of_chars = [image]
    char_img_converted_to_in_sample = []
    char_img_heights = []
    char_img_widths = []

    for char_img in array_of_chars:
        char_img_heights.append(len(char_img))
        char_img_widths.append(len(char_img[0]))

        # convert to a numpy array
        char_img = np.array(char_img, dtype='float32')

        # resize the image, reshape, make prediction
        char_img = cv2.resize(char_img, (IMG_SIZE, IMG_SIZE))

        # normalize the image
        char_img /= 255
        char_img = char_img.reshape(-1, IMG_SIZE, IMG_SIZE, 1)
        char_img_converted_to_in_sample.append(char_img)
    return char_img_converted_to_in_sample, 1, char_img_heights, char_img_widths



image_number = 1
while os.path.isfile(f"{TEST_DIR}/{image_number}.png"):
    try:
        final_images, space_location, char_img_heights, char_img_widths = prepare(image_number)
        final_prediction = []
        for idx, img in enumerate(final_images):

            char_prediction_1 = make_prediction(model_1, img)
            char_prediction_2 = make_prediction(model_2, img)
            char_prediction_3 = make_prediction(model_3, img)
            char_prediction_4 = make_prediction(model_4, img)
            char_prediction_5 = make_prediction(model_5, img)

            print('\n', char_prediction_1, char_prediction_2, char_prediction_3, char_prediction_4, char_prediction_5)

            # Combined prediction
            final_char_prediction = model_jury_ruling(char_prediction_1, char_prediction_2, char_prediction_3,
                                                      char_prediction_4, char_prediction_5)

            im = Image.open(f"{TEST_DIR}/{image_number}.png")
            LOWER_CASE = im.height * .50
            TALL_UNIFORM_LC = im.height * .70

            tall_uniform_lc_letters = 'klpy'

            if final_char_prediction.isnumeric() == False and final_char_prediction.lower() not in class_mapping and TEST_CASE_SMALL == True:
                if final_char_prediction.lower() in tall_uniform_lc_letters:
                    if char_img_heights[idx] < TALL_UNIFORM_LC:
                        final_char_prediction = final_char_prediction.lower()
                elif char_img_heights[idx] < LOWER_CASE:
                    final_char_prediction = final_char_prediction.lower()

            if final_char_prediction == '0' and char_img_heights[idx] < LOWER_CASE and TEST_CASE_SMALL == True:
                final_char_prediction = 'o'

            final_prediction.append(final_char_prediction)
            print('Final Prediction: ', final_prediction)
            plt.imshow(img[0], cmap=plt.cm.binary)
            plt.title(f"Testing Number: {image_number} \n Prediction: {final_char_prediction}")
            plt.show()

    except:
        print("Error reading image! Proceeding with next image...")
    finally:
        image_number += 1


train_dataset.py
import os
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import tensorflow as tf
import pickle

# defining the directories
DATA_TRAIN_DIR = 'dataset/emnist-balanced-train.csv'  # directory of the dataset used to TRAIN the model
DATA_TEST_DIR = 'dataset/emnist-balanced-test.csv'  # directory of the dataset used to TEST the model
DATA_MAP_DIR = 'dataset/emnist-balanced-mapping.txt'  # directory of the dataset used for the label mapping
GENERATED_PICKLE_DIR = 'generated_pickle'

# defining class related variables
class_mapping = []
class_length = 0

# defining image size
HEIGHT = 28
WIDTH = 28

# defining the train and test dataset with header none
train_pd = pd.read_csv(DATA_TRAIN_DIR, header=None)
test_pd = pd.read_csv(DATA_TEST_DIR, header=None)
map_pd = pd.read_csv(DATA_MAP_DIR, delimiter=' ', header=None, usecols=[1]).squeeze(True)

# remapping the map labels to local variable class_mapping
for num in range(len(map_pd)):
    class_mapping.append(chr(map_pd[num]))

# determining the length of class map
class_length = len(class_mapping)


# generate an image base on the test number
def covert_training_data(dataset, image_number):
    test_row_num_img = dataset.values[image_number, 1:]
    reshape_img = test_row_num_img.reshape(HEIGHT, WIDTH)
    reshape_img = np.transpose(reshape_img, axes=[1, 0])
    return reshape_img


# return the label of the row
def get_char(dataset, row):
    return class_mapping[dataset.values[row, 0]]


# return the generated dataset with proper sets
def generate_proper_datasets(dataset):
    result = []
    for i in range(len(dataset)):
        result.append(covert_training_data(dataset, i))
    return np.asarray(result)


# return the normalized dataset
def normalize_dataset(dataset):
    result = dataset.astype('float32')
    return result / 255


# generate the proper dataset
train_x = generate_proper_datasets(train_pd)
test_x = generate_proper_datasets(test_pd)

# normalize the train and test data
train_x = normalize_dataset(train_x)
test_x = normalize_dataset(test_x)

# creating the answers
train_y = train_pd.iloc[:, 0]
test_y = test_pd.iloc[:, 0]

# for later
train_y_1D = train_y
test_y_1D = test_y

train_x_3D = train_x
test_x_3D = test_x


# prepare datasets for ML algorithm
train_y = tf.keras.utils.to_categorical(train_y, class_length)
test_y = tf.keras.utils.to_categorical(test_y, class_length)


# generate pickle for the dataset
def generate_pickle(title, dataset):
    print(f'generating {title}.pickle')
    pickle_out = open(f"{GENERATED_PICKLE_DIR}/{title}.pickle", 'wb')
    pickle.dump(dataset, pickle_out)
    pickle_out.close()


generate_pickle('train_x', train_x)
generate_pickle('test_x', test_x)
generate_pickle('train_y', train_y)
generate_pickle('test_y', test_y)





model_training/model_permutation.py
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation, Flatten
from tensorflow.keras.layers import Conv2D, MaxPooling2D
from tensorflow.keras.callbacks import TensorBoard
from sklearn.model_selection import train_test_split
import pickle
import time

HEIGHT = 28
WIDTH = 28
class_num = 47

train_x = pickle.load(open("../generated_pickle/train_x.pickle", "rb"))
train_y = pickle.load(open("../generated_pickle/train_y.pickle", "rb"))

test_x = pickle.load(open("../generated_pickle/test_x.pickle", "rb"))
test_y = pickle.load(open("../generated_pickle/test_y.pickle", "rb"))

train_x, val_x, train_y, val_y = train_test_split(train_x, train_y, test_size=0.10, random_state=7)

# 1st run: create the following permutations of models
dense_layers = [2]
layer_sizes = [64]
conv_layers = [2]

for dense_layer in dense_layers:
    for layer_size in layer_sizes:
        for conv_layer in conv_layers:

            OPERATION_NAME = f"{conv_layer}-conv-{layer_size}-nodes-{dense_layer}-dense-{int(time.time())}"

            tensorboard = TensorBoard(log_dir=f"logs/{OPERATION_NAME}")

            print(OPERATION_NAME)

            model = Sequential()

            # Begin with Conv2D Layer
            model.add(Conv2D(layer_size, (3, 3), input_shape=(HEIGHT, WIDTH, 1)))
            model.add(Activation('relu'))
            model.add(MaxPooling2D(pool_size=(2, 2)))

            # For Extra Conv2D Layers
            for l in range(conv_layer - 1):
                model.add(Conv2D(layer_size, (3, 3)))
                model.add(Activation('relu'))
                model.add(MaxPooling2D(pool_size=(2, 2)))

            # Flatten the Model to have a single stream
            model.add(Flatten())

            # For Dense Layer
            for l in range(dense_layer):
                model.add(Dense(layer_size))
                model.add(Activation('relu'))
                model.add(Dropout(0.2))

            # Output Layer
            model.add(Dense(units=class_num, activation='softmax'))

            # Train Model
            model.compile(
                optimizer='adam',
                loss='categorical_crossentropy',
                metrics=['accuracy']
            )

            model.summary()

            model.fit(
                train_x,
                train_y,
                epochs=10,
                batch_size=32,
                validation_data=(val_x, val_y),
                callbacks=[tensorboard]
            )

            model.save('models/model_5.h5')


predict.py
import tensorflow as tf
import numpy as np
import pandas as pd
import pickle
from sklearn.model_selection import train_test_split
import cv2
import os
import matplotlib.pyplot as plt
from PIL import Image


model_1 = tf.keras.models.load_model('models/model_1.h5')
model_2 = tf.keras.models.load_model('models/model_2.h5')
model_3 = tf.keras.models.load_model('models/model_3.h5')
model_4 = tf.keras.models.load_model('models/model_4.h5')
model_5 = tf.keras.models.load_model('models/model_5.h5')

DATA_MAP_DIR = 'dataset/emnist-balanced-mapping.txt'  # directory of the dataset used for the label mapping

map_pd = pd.read_csv(DATA_MAP_DIR, delimiter=' ', header=None, usecols=[1]).squeeze(True)

class_mapping = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabdefghnqrt'
class_length = 0

# # remapping the map labels to local variable class_mapping
# for num in range(len(map_pd)):
#     class_mapping.append(chr(map_pd[num]))

# determining the length of class map
class_length = len(class_mapping)

train_x = pickle.load(open("generated_pickle/train_x.pickle", "rb"))
train_y = pickle.load(open("generated_pickle/train_y.pickle", "rb"))

test_x = pickle.load(open("generated_pickle/test_x.pickle", "rb"))
test_y = pickle.load(open("generated_pickle/test_y.pickle", "rb"))

train_x, val_x, train_y, val_y = train_test_split(train_x, train_y, test_size=0.10, random_state=7)

def make_prediction(model, img):
    prediction = model.predict(img)
    idx_prediction = np.argmax(prediction[0])
    return class_mapping[idx_prediction]

def model_jury_ruling(*argv):
    all_predictions = []
    for arg in argv:
        all_predictions.append(arg)

    hash = {}

    for prediction in all_predictions:
        if prediction not in hash:
            hash[prediction] = 1
        else:
            hash[prediction] += 1

    return max(hash, key=hash.get)


def take_all_prediction(image):
    predictions = []
    models = [model_1, model_2, model_3, model_4, model_5]
    for model in models:
        predicted = make_prediction(model, image)
        predictions.append(predicted)
    return predictions


def prepare(image_number):
    IMG_SIZE = 28
    img = cv2.imread('test/{}.png'.format(image_number), cv2.IMREAD_GRAYSCALE)
    img = np.invert(np.array([img]))
    array_of_chars = [img]
    char_img_converted_to_in_sample = []
    char_img_heights = []
    char_img_widths = []

    for char_img in array_of_chars:
        # trim off all excess pixels and center the char_img up
        # char_img = center_image(char_img)
        char_img_heights.append(len(char_img))
        char_img_widths.append(len(char_img[0]))

        # we will now begin padding
        # convert to a numpy array
        char_img = np.array(char_img, dtype='float32')

        # resize the image, reshape, make prediction
        char_img = cv2.resize(char_img, (IMG_SIZE, IMG_SIZE))
        char_img /= 255

        # plt.imshow(char_img, cmap=plt.cm.binary)
        # plt.show()

        char_img = char_img.reshape(-1, IMG_SIZE, IMG_SIZE, 1)
        char_img_converted_to_in_sample.append(char_img)
    return char_img_converted_to_in_sample, 1, char_img_heights, char_img_widths


image_number = 1
while os.path.isfile('test/{}.png'.format(image_number)):
    try:
        final_images, space_location, char_img_heights, char_img_widths = prepare(image_number)
        final_prediction = []
        for idx, img in enumerate(final_images):

            char_prediction_1 = make_prediction(model_1, img)
            char_prediction_2 = make_prediction(model_2, img)
            char_prediction_3 = make_prediction(model_3, img)
            char_prediction_4 = make_prediction(model_4, img)
            char_prediction_5 = make_prediction(model_5, img)

            print('\n', char_prediction_1, char_prediction_2, char_prediction_3, char_prediction_4, char_prediction_5)

            # Combined prediction
            final_char_prediction = model_jury_ruling(char_prediction_1, char_prediction_2, char_prediction_3,
                                                      char_prediction_4, char_prediction_5)
            im = Image.open(f"test/{image_number}.png")
            LOWER_CASE = im.height * .50
            tall_uniform_lc_letters = 'klpy'
            TALL_UNIFORM_LC = im.height * .70

            if final_char_prediction.isnumeric() == False and final_char_prediction.lower() not in class_mapping:
                if final_char_prediction.lower() in tall_uniform_lc_letters:
                    if char_img_heights[idx] < TALL_UNIFORM_LC:
                        final_char_prediction = final_char_prediction.lower()
                elif char_img_heights[idx] < LOWER_CASE:
                    final_char_prediction = final_char_prediction.lower()

            # Typically, "zeroes" (0) are fairly large, we would typically rather have "o" instead if a 0 if a user makes them small
            if final_char_prediction == '0' and char_img_heights[idx] < LOWER_CASE:
                final_char_prediction = 'o'


            final_prediction.append(final_char_prediction)
            print('Final Prediction: ', final_prediction)
            plt.imshow(img[0], cmap=plt.cm.binary)
            plt.title(f"Prediction: {final_char_prediction}")
            plt.show()



    #
    except:
        print("Error reading image! Proceeding with next image...")
    finally:
        image_number += 1